src/pullfwds/qr.jl

```diff
@@ -0,0 +1,69 @@
+function qr_pullfwd!(dA, A, QR, dQR; tol::Real=MatrixAlgebraKit.default_pullback_gaugetol(QR[2]), rank_atol::Real=tol, gauge_atol::Real=tol)
+    Q, R  = QR
+    m     = size(Q, 1)
+    n     = size(R, 2)
+    minmn = min(m, n)
+    Rd    = diagview(R)
+    p     = findlast(>=(rank_atol) ∘ abs, Rd)
+
+    m1 = p
+    m2 = minmn - p
+    m3 = m - minmn
+    n1 = p
+    n2 = n - p
+
+    Q1  = view(Q, 1:m, 1:m1) # full rank portion
+    Q2  = view(Q, 1:m, m1+1:m2+m1)
+    R11 = view(R, 1:m1, 1:n1)
+    R12 = view(R, 1:m1, n1+1:n)
+
+    dA1 = view(dA, 1:m, 1:n1)
+    dA2 = view(dA, 1:m, (n1 + 1):n)
+
+    dQ, dR = dQR
+    dQ1    = view(dQ, 1:m, 1:m1)
+    dQ2    = view(dQ, 1:m, m1+1:m2+m1)
+    dQ3    = m1+m2+1 < size(dQ, 2) ? view(dQ, 1:m, m1+m2+1:size(dQ,2)) : similar(dQ, eltype(dQ), (0, 0))
```

Comment:
Do you mean `m1+m2 < size(dQ, 2)` or `m1+m2+1 <= size(dQ, 2)`?

Also, does just using `view(dQ, 1:m, m1+m2+1:size(dQ,2))` always not just work? In the case that `m1+m2 == size(dQ, 2)`, we automatically get a view with `size(dQ3, 2)=0`, no ?